### ISSUE TITLE
[Enhancement] Use regex library permitting lookaround and lookaheads

### ISSUE THREAD

**Issue by apmclean** [aws-cloudformation/cloudformation-guard#114]
When evaluating Statement entries in a policy, permitting lookaround and lookaheads in the regular expression syntax can be helpful to verify a statement.

Given the following S3 Bucket Policy snippet:

```yaml
      PolicyDocument:
        Statement:
          - Action: s3:PutObject
            Condition:
              StringEquals:
                s3:x-amz-server-side-encryption: AES256
            Effect: Deny
            Principal: "*"
            Resource: [ self ]
          - Action: s3:*
            Condition:
              Bool:
                aws:SecureTransport: "false"
            Effect: Deny
            Principal: "*"
            Resource: [ self ]
```

I would like a guard rule that confirms these policies are correctly formed under the following circumstances:
- The statement array can be any length (in practice our templates have 20 statement entries).
- The index number of these policies cannot be known in advance.  ie: they may be at position 0, and 2 during one evaluation then position 2, and 8 during another.
- The key order within the statement can be any arbitrary order.  ie: `{"Action": "s3:*", "Effect": "Deny"}` is treated the same as `{"Effect": "Deny", "Action": "s3:*"}` as any key order is considered valid CloudFormation.

Using PCRE with lookarounds I'm able to solve this using something like this:
`(?=.*Action)(?=.*s3:\*)(?=.*SecureTransport)(?=.*false)(?=.*Deny)`

I could then use this in a statement wildcard rule like this:
`AWS::S3::BucketPolicy PolicyDocument.Statement.* == /(?=.*Action)(?=.*s3:\*)(?=.*SecureTransport)(?=.*false)(?=.*Deny)/`

Currently the Rust Regex engine used by the project doesn't support lookaround/lookaheads: `Its syntax is similar to Perl-style regular expressions, but lacks a few features like look around and backreferences.`

However the 'Fancy' one does: https://docs.rs/fancy-regex/0.4.1/fancy_regex/

Feature request is to switch the project to use fancy_regex to enable this style of rule parsing.

**Comment by jotompki** [comment]
The need for these lookahead regexes arise when comparing serialized json objects to strings/regular expressions. This is a limitation of the current guard tool. In #115, we are detailing out some changes that should make accomplishing the checks you want a bit more straightforward. Instead of using a regular expression, you will be able to just write checks on the properties. Something like (syntax still WIP):
```
AWS::S3::Bucket {
   let statements = PolicyDocument.Statement
   %statements[*].Action == "s3:*"
   %statements[*].SecureTransport == false
   %statements[*].Effect == "Deny"
 }
```

I think that the use of these lookaround and lookaheads comes about because of the limited equality comparison for objects. We want customers to be able to specify this simply and not using regular expressions.

**Comment by dchakrav-github** [comment]
@apmclean this is addressed in 2.0, we compare value objects and not string comparisons which makes evaluations not order dependent on Key/Value pairs for Maps or index dependent for lists that 1.0 did. As @jotompki pointed out you can navigate the JSON structure of a policy and have it evaluated correctly. You can also conditionally evaluate using `when` condition checks, which can also check for multiple conditions to be met for assertions.

A couple of corrections on @jotompki, i think he meant type is `AWS::IAM::Role` or `AWS::IAM::BucketPolicy` not `AWS::S3::Bucket`. Here are a couple of explanations for usage 

1. You wanted all IAM policy document statements or S3 bucket policy statements or Managed Policy to be compared for asserting that `NotAction` or `NotPrincipal` is NOT used

```
let policy_documents = Resources.*[ 
  Type == 'AWS::IAM::Policy'         OR
  Type == 'AWS::S3::BucketPolicy'    OR 
  Type == 'AWS::IAM::ManagedPolicy'
].PolicyDocument

rule disallow_not_action_principal_check WHEN %policy_documents NOT EMPTY {
  # 
  # Ensure that none of the Statement for the policy DOES NOT contain NotAction or NotPrincipal
  # 
  %policy_documents.Statement[*] {
    NotAction NOT EXISTS
    NotPrinciapl NOT EXISTS
  }
}
```

2. You want to assert all `s3` actions MUST use secure Transport 

```
rule ensure_secure_s3_transport WHEN %policy_documents NOT EMPTY  {
  # 
  # Make sure all s3:* actions only use secure transport when specified
  #
  %policy_documents.Statement[*] {
    WHEN Effect == 'Allow'
         SOME Action[*] == /s3.*/ # at-least-one action equals s3 based
    {
      Condition.Bool.'aws:SecureTransport' == "true" # we do not auto-convert to keep it compliant with what IAM expects as values
    }
  }
}
```

There are several more checks for company-wide policy that can be enforced.

Here is the complete rule 
```
let policy_documents = Resources.*[ 
  Type == 'AWS::IAM::Policy'         OR
  Type == 'AWS::S3::BucketPolicy'    OR 
  Type == 'AWS::IAM::ManagedPolicy'
].Properties.PolicyDocument

rule disallow_not_action_principal_check WHEN %policy_documents NOT EMPTY {
  # 
  # Ensure that none of the Statement for the policy DOES NOT contain NotAction or NotPrincipal
  # 
  %policy_documents.Statement[*] {
    NotAction NOT EXISTS
    NotPrinciapl NOT EXISTS
  }
}

rule ensure_secure_s3_transport WHEN %policy_documents NOT EMPTY  {
  #
  # Make sure all s3:* actions only use secure transport when specified
  #
  %policy_documents.Statement[*] {
    WHEN Effect == 'Allow'
         SOME Action[*] == /s3.*/ # at-least-one action equals s3 based
    {
      Condition.Bool.'aws:SecureTransport' == "true" # we do not auto-convert to keep it compliant with what IAM expects as values
    }
  }
}
```  

Here is the sample unit test file
```YAML
---
- input:
    Resources:
      policy:
        Type: 'AWS::IAM::Policy'
        Properties:
          PolicyDocument:
            Statement:
              - Effect: Allow
                Action: s3:PutObject
                Condition:
                  Bool:
                    aws:SecureTransport: "true"
  expectations:
    rules:
      disallow_not_action_principal_check: PASS
      ensure_secure_s3_transport: PASS

- input:
    Resources:
      policy:
        Type: 'AWS::IAM::Policy'
        Properties:
          PolicyDocument:
            Statement:
              - Effect: Allow
                Action: s3:PutObject
                Resource: self
                Condition:
                  Bool:
                    aws:SecureTransport: "true"
              - Effect: Allow
                Action: 'ec2:*'
                Resource: '*'
  expectations:
    rules:
      disallow_not_action_principal_check: PASS
      ensure_secure_s3_transport: PASS

- input:
    Resources:
      policy:
        Type: 'AWS::IAM::Policy'
        Properties:
          PolicyDocument:
            Statement:
              - Effect: Allow
                Action: 'ec2:*'
                Resource: '*'
  expectations:
    rules:
      disallow_not_action_principal_check: PASS
      ensure_secure_s3_transport: SKIP

- input:
    Resources:
      policy:
        Type: 'AWS::IAM::Policy'
        Properties:
          PolicyDocument:
            Statement:
              - Effect: Allow
                NotAction: 'ec2:*'
                Resource: '*'
  expectations:
    rules:
      disallow_not_action_principal_check: FAIL
      ensure_secure_s3_transport: SKIP

- input:
    Resources:
      policy:
        Type: 'AWS::IAM::Policy'
        Properties:
          PolicyDocument:
            Statement:
              - Effect: Allow
                NotAction: 'ec2:*'
                Resource: '*'
              - Effect: Allow
                Action: 's3:*'
                Resource: self
                Condition:
                  Bool:
                    aws:SecureTransport: "true"
  expectations:
    rules:
      disallow_not_action_principal_check: FAIL
      ensure_secure_s3_transport: PASS

- input:
    Resources:
      policy:
        Type: 'AWS::IAM::Policy'
        Properties:
          PolicyDocument:
            Statement:
              - Effect: Allow
                NotAction: 'ec2:*'
                Resource: '*'
              - Effect: Allow
                Action: 's3:*'
                Resource: self
  expectations:
    rules:
      disallow_not_action_principal_check: FAIL
      ensure_secure_s3_transport: FAIL

- input:
    Resources:
      policy:
        Type: 'AWS::IAM::ManagedPolicy'
        Properties:
          PolicyDocument:
            Statement:
              - Effect: Allow
                NotAction: 'ec2:*'
                Resource: '*'
              - Effect: Allow
                Action: 's3:*'
                Resource: self
                Condition:
                  Bool:
                    aws:SecureTransport: "false"
  expectations:
    rules:
      disallow_not_action_principal_check: FAIL
      ensure_secure_s3_transport: FAIL
```

**Comment by priyap286** [comment]
Closing this issue as there are no more open questions to the best of my knowledge. Please reopen this issue if you think there are missed open issues on this topic.
